apps/crl.c

```diff
@@ -231,7 +236,150 @@ int crl_main(int argc, char **argv)
+        counts = OPENSSL_malloc(ext_total * sizeof(*counts));
+        if (counts == NULL) {
+            BIO_printf(bio_err,
+                       "Out of memory counting extensions\n");
+        } else {
+            for (i = 0; i < ext_total; i++) {
+                ext = sk_X509_EXTENSION_value(exts, i);
+                nid = OBJ_obj2nid(X509_EXTENSION_get_object(ext));
+                for (j = 0; j < distinct; j++) {
+                    if (counts[j].nid == nid) {
+                        counts[j].count++;
+                        break;
+                    }
+                }
+                if (j == distinct) {
+                    counts[distinct].nid   = nid;
+                    counts[distinct].count = 1;
+                    distinct++;
+                }
+            }
+            for (i = 0; i < distinct; i++) {
+                name = OBJ_nid2sn(counts[i].nid);
+                if (name == NULL)
+                    name = OBJ_nid2ln(counts[i].nid);
+                BIO_printf(bio_err,
+                           "Extension %-30s (NID %3d): %d\n",
+                           name,
+                           counts[i].nid,
+                           counts[i].count);
+            }
+            /*
+            * Enforce RFC 5280 §5.1.2.7 / §5.2 rules:
+            *  - Authority Key Identifier: exactly one
+            *  - CRL Number: xactly one
+            *  - IssuerAltName, DeltaCRLIndicator,
+            *    IssuingDistributionPoint, FreshestCRL, AIA:
+            *    at most one each
+            *  - Any critical unknown extension -> error
+            */
+            for (i = 0; i < distinct; i++) {
+                nid = counts[i].nid;
+                cnt = counts[i].count;
+                sn = OBJ_nid2sn(nid);
+                if (sn == NULL)
+                    sn = OBJ_nid2ln(nid);
+                switch (nid) {
+                case NID_authority_key_identifier:
+                case NID_crl_number:
+                    if (cnt != 1) {
+                        BIO_printf(bio_err,
+                                   "CRL extension %s must appear "
+                                   "exactly once (found %d)\n",
+                                   sn, cnt);
+                        ret = 1;
+                        OPENSSL_free(counts);
+                        goto end;
+                    }
+                    break;
+                case NID_issuer_alt_name:
+                case NID_delta_crl:
+                case NID_issuing_distribution_point:
+                case NID_freshest_crl:
+                case NID_info_access:
+                    if (cnt > 1) {
+                        BIO_printf(bio_err,
+                                   "CRL extension %s must appear "
+                                   "at most once (found %d)\n",
+                                   sn, cnt);
+                        ret = 1;
+                        OPENSSL_free(counts);
```

Comment:
This should be a warning, not a failure to parse the CRL, perhaps with a flag that makes the warning fatal?. What is the purpose of refusing the parse the CRL?
Indeed extensions MUST not appear more than once in a certificate or CRL, but the `crl(1)` utility should not be where this is enforced. 
When a CRL has problems, a user should be able to inspect it with this utility, and failing to parse is likely counterproductive.

The place to reject ambiguous CRLs (as a result of multiple instances of an extension) is when the CRL is used as part of certificate validation. In that context we need to exercise judgment about which issues are relevant to enforce when verifying rather than advice to CAs to ensure interoperability.